tls/s2n_early_data_io.c

```diff
@@ -117,3 +119,135 @@ S2N_RESULT s2n_early_data_validate_recv(struct s2n_connection *conn)
+    if (negotiate_result < S2N_SUCCESS) {
+        RESULT_ENSURE_EQ(s2n_error_get_type(negotiate_error), S2N_ERR_T_BLOCKED);
+        if (negotiate_blocked == S2N_BLOCKED_ON_EARLY_DATA) {
+            return S2N_RESULT_OK;
+        } else if (s2n_early_data_can_continue(conn)) {
+            *blocked = negotiate_blocked;
+            RESULT_BAIL(negotiate_error);
+        } else {
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+int s2n_send_early_data(struct s2n_connection *conn, const uint8_t *data, ssize_t data_len,
+        ssize_t *data_sent, s2n_blocked_status *blocked)
+{
+    /* Calling this method indicates that we expect early data. */
+    POSIX_GUARD(s2n_connection_set_early_data_expected(conn));
+
+    s2n_result result = s2n_send_early_data_impl(conn, data, data_len, data_sent, blocked);
+
+    /* Unless s2n_send_early_data is called again (undoing this), we are done sending early data.
+     * If s2n_negotiate is called next, we could send the EndOfEarlyData message. */
+    POSIX_GUARD(s2n_connection_set_end_of_early_data(conn));
+
+    POSIX_GUARD_RESULT(result);
+    return S2N_SUCCESS;
+}
+
+S2N_RESULT s2n_recv_early_data_impl(struct s2n_connection *conn, uint8_t *data, ssize_t max_data_len,
+        ssize_t *data_received, s2n_blocked_status *blocked)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(blocked);
+    *blocked = S2N_NOT_BLOCKED;
+    RESULT_ENSURE_REF(data_received);
+    *data_received = 0;
+
+    RESULT_ENSURE(conn->mode == S2N_SERVER, S2N_ERR_CLIENT_MODE);
+
+    if (!s2n_early_data_can_continue(conn)) {
+        return S2N_RESULT_OK;
+    }
+
+    while(s2n_negotiate(conn, blocked) < S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            return S2N_RESULT_ERROR;
+        } else if (max_data_len <= *data_received) {
+            return S2N_RESULT_ERROR;
+        } else if (*blocked != S2N_BLOCKED_ON_EARLY_DATA) {
+            if (s2n_early_data_can_continue(conn)) {
+                return S2N_RESULT_ERROR;
+            } else {
+                *blocked = S2N_NOT_BLOCKED;
+                return S2N_RESULT_OK;
+            }
+        }
+
+        ssize_t recv_result = s2n_recv(conn, data + *data_received,
+                max_data_len - *data_received, blocked);
+        RESULT_GUARD_POSIX(recv_result);
+        *data_received += recv_result;
+    }
+    return S2N_RESULT_OK;
+}
+
+int s2n_recv_early_data(struct s2n_connection *conn, uint8_t *data, ssize_t max_data_len,
+        ssize_t *data_received, s2n_blocked_status *blocked)
+{
+    /* Calling this method indicates that we expect early data. */
+    POSIX_GUARD(s2n_connection_set_early_data_expected(conn));
+
+    s2n_result result = s2n_recv_early_data_impl(conn, data, max_data_len, data_received, blocked);
+
+    /* Unless s2n_recv_early_data is called again (undoing this), we are done accepting early data. */
```

Comment:
What happens if we call s2n_negotiate with early data remaining to receive?

Reply:
https://github.com/aws/s2n-tls/blob/61b996625635847048449ffc6b2bf3f2969cd181/tests/unit/s2n_early_data_io_api_test.c#L342-L348